src/components/tl/ucp/tl_ucp_ep.c

```diff
@@ -81,13 +95,13 @@ void ucc_tl_ucp_close_eps(ucc_tl_ucp_context_t *ctx)
 
      param.op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS;
      param.flags        = 0; // 0 means FLUSH
-     ep                 = get_next_ep_to_close(ctx, &i);
+     ep                 = get_next_ep_to_close(ep_hash, eps, ctx, &i);
      while (ep) {
          close_req = ucp_ep_close_nbx(ep, &param);
 
          if (UCS_PTR_IS_PTR(close_req)) {
              do {
-                 ucp_worker_progress(ctx->ucp_worker);
+                 ucp_worker_progress(worker);
```

Comment:
i think here we should be progressing both workers if 2 are available, otherwise we can get a deadlock
This in general applies to all do {} while blocking loops where we progress ucp worker

Reply:
done, see [TL/UCP: fix deadlock w/ service & default workers](https://github.com/openucx/ucc/pull/560/commits/67f6a2b1c6cd318c0248da669d379565aa5f0618)